Migrate mongoose plugin to opentelemetryjs-contrib

wdalmut opened the pull request:
As mentioned in #18 i propose the mongoose plugin migration to the opentelemetryjs-contrib repository.

Currently this plugin is available in the opentelemetry registry: https://opentelemetry.io/registry/instrumentation-js-mongoose/

wdalmut commented:
i have to fix all issues i just need time to close everything
/cc @vmarchaud

vmarchaud commented:
@wdalmut No worries, take the time you need. If you need any help tell me

wdalmut commented on plugins/node/opentelemetry-plugin-mongoose/.travis.yml:
```diff
@@ -0,0 +1,14 @@
+language: node_js
```
i have to drop this...

wdalmut commented on plugins/node/opentelemetry-plugin-mongoose/LICENSE:
```diff
@@ -0,0 +1,9 @@
+Copyright 2020 Walter Dal Mut
```
Need to update this file

wdalmut commented:
@vmarchaud i think this could be my first proposal...

vmarchaud commented on plugins/node/opentelemetry-plugin-mongoose/src/mongoose.ts:
```diff
@@ -0,0 +1,279 @@
+
+// when mongoose functions are called, we store the original call context
+// and then set it as the parent for the spans created by Query/Aggregate exec()
+// calls. this bypass the unlinked spans issue on thenables await operations (issue #29)
```
i think you need to update the reference to the issue

vmarchaud commented on plugins/node/opentelemetry-plugin-mongoose/src/mongoose.ts:
```diff
@@ -0,0 +1,279 @@
+  // on the Aggregate object to capture the context on, so we patch
+  // the aggregate of Model, and set the context on the Aggregate object
+  private patchModelAggregate() {
+    const thisPlugin = this;
```
we generally call this reference `plugin`

vmarchaud commented on plugins/node/opentelemetry-plugin-mongoose/src/mongoose.ts:
```diff
@@ -0,0 +1,279 @@
+        if (fn instanceof Function) {
+          return originalOnModelFunction.apply(this, [
+            options,
+            (err: Error, product: any) => {
```
what is `product` here ?

vmarchaud commented on plugins/node/opentelemetry-plugin-mongoose/jasmine.json:
```diff
@@ -0,0 +1,12 @@
+{
```
we currently use mocha for all of our tests so i think those tests would need to be converted (not sure if its a hard requirement though)

vmarchaud commented on plugins/node/opentelemetry-plugin-mongoose/package.json:
```diff
@@ -0,0 +1,70 @@
+    "README.md"
+  ],
+  "dependencies": {
+    "@opentelemetry/api": "^0.6.0",
```
could you bump tose versions to latest ?

vmarchaud commented on plugins/node/opentelemetry-plugin-mongoose/README.md:
```diff
@@ -0,0 +1,23 @@
+# OpenTelemetry Mongoose Plugin
```
This need to match the style of ours readmes, example with mongodb: https://github.com/open-telemetry/opentelemetry-js-contrib/blob/master/plugins/node/opentelemetry-plugin-mongodb/README.md

vmarchaud commented on plugins/node/opentelemetry-plugin-mongoose/src/mongoose.ts:
```diff
@@ -0,0 +1,279 @@
+// when mongoose functions are called, we store the original call context
+// and then set it as the parent for the spans created by Query/Aggregate exec()
+// calls. this bypass the unlinked spans issue on thenables await operations (issue #29)
+export const _STORED_PARENT_SPAN: unique symbol = Symbol('stored-parent-span');
```
Currently exporting here means end-user could have access to the symbol, i think you should move it inside the utils and importing it here

wdalmut commented:
I have fixed all notes and moved the whole test suite to mocha

//cc @vmarchaud

wdalmut commented:
> I have fixed all notes and moved the whole test suite to mocha
> 
> //cc @vmarchaud

i miss some comment in the review but i think i can solve everything today...

wdalmut reviewed (commented):
All comments now are included

CyberHippo commented:
@wdalmut Thank you for your great work! Could you please merge it?

wdalmut commented:
@vmarchaud do you have updates on this?

vmarchaud commented:
@wdalmut Well i was fine with merging this at the time, but it seems we have a problem for reviewing PR on this repo. In the mean time we updated the way plugin are implemented to fix few issues (see [the new base package](https://github.com/open-telemetry/opentelemetry-js/tree/master/packages/opentelemetry-instrumentation)).

However the new system is not yet finished (we dont have a loader for those) so i think it would be better to wait, update the plugin to use the new system and then merge it on the contrib repo.

obecny commented:
@dyladan @vmarchaud what should be done with this plugin as it is outdated anyway ?
1. update to latest, but don't convert yet to instrumentation, once it is done, then check it and finally merge, then convert it to instrumentation
2. refactor this to be an instrumentation, update to latest, review, merge,
3. any other options ?

vmarchaud commented:
I woud prefer to not merge it so we can get rid of all plugins to release 1.0 of SDK. So i'll be in favor of closing this one and create an issue to port it to instrumentation.

vmarchaud commented:
I'm going to close this as we are deprecating the `BasePlugin` system, if people are looking for an instrumentation you should checkout https://github.com/aspecto-io/opentelemetry-ext-js/tree/master/packages/instrumentation-mongoose
